### PR TITLE
BXC-3335 - Support resumption of export operations

### PR DESCRIPTION
**Pull request by bbpennel**
https://jira.lib.unc.edu/browse/BXC-3335

* Export operations will attempt to pick up after the last page listed or exported in order to support export of large collections in the case the process is interrupted
    * In order to totally restart an export, the `-f` flag will need to be provided if an export was previously started or completed.
    * The user will be alerted if an export is resuming, or if they are being prevented from starting a new one.
    * Since there are a few major stages to the export process (counting objects, listing ids, exporting metadata), implementation is a simple state machine, with state persisted to a json file, plus a flat text file for recording object ids.
* Updates to related services to bring them more in line with services developer later, and because the list of parameters was getting long: setting the project as an instant variable, using an Options mixin to more easily pass in flags to services.
* Added a commandline option for setting page size for listing of objects, mostly to support testing

### CODE REVIEW

File: src/main/java/edu/unc/lib/boxc/migration/cdm/services/CdmExportService.java [hunk 1]
```diff
@@ -49,68 +51,95 @@
...
      * Export all records from CDM for the provided project
-     * @param project
+     * @param options
      * @throws IOException
      */
-    public void exportAll(MigrationProject project) throws IOException {
+    public void exportAll(CdmExportOptions options) throws IOException {
         // Generate body of export request using the list of fields configure for export
         cdmFieldService.validateFieldsFile(project);
+        initializeExportDir(project);
```
[the remark this answers is not in the document]
Comment (bbpennel) [reply src/main/java/edu/unc/lib/boxc/migration/cdm/services/CdmExportService.java:70]: It is getting injected like some of the other dependencies. In a non-test context, it gets initialized here:
https://github.com/UNC-Libraries/cdm-to-boxc-migration-util/pull/22/files/206095afe0464df90b19a8770308df898a59b2ae#diff-c4f2dd07f64d6e7a96a3049af337e888219b5c454cabd9eac17ec92235d6b9f1R78

And injected into CdmExportService here:
https://github.com/UNC-Libraries/cdm-to-boxc-migration-util/pull/22/files/206095afe0464df90b19a8770308df898a59b2ae#diff-c4f2dd07f64d6e7a96a3049af337e888219b5c454cabd9eac17ec92235d6b9f1R98

These steps both happen before `exportAll` is called

File: src/main/java/edu/unc/lib/boxc/migration/cdm/services/CdmExportService.java [hunk 2]
```diff
@@ -49,68 +51,95 @@
...
-                            + "): " + IOUtils.toString(resp.getEntity().getContent(), ISO_8859_1));
+        if (exportStateService.inStateOrNotResuming(ProgressState.LISTING_COMPLETED)) {
+            exportStateService.transitionToExporting(pageSize);
+        }
+        if (exportStateService.inStateOrNotResuming(ProgressState.EXPORTING)) {
+            int exportPage = 0;
+            for (List<String> chunk : chunks) {
+                // Name each exported page
+                exportPage++;
+                if (exportStateService.isResuming()
```
Comment (bbpennel) [inline src/main/java/edu/unc/lib/boxc/migration/cdm/services/CdmExportService.java:90]: Make it clear what's happening here, documentation or a helper method

File: src/main/java/edu/unc/lib/boxc/migration/cdm/services/export/ExportState.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+import java.time.Instant;
+
+/**
+ * State of a CDM export operation
+ *
+ * @author bbpennel
+ */
+public class ExportState {
+    private Instant startTime;
```
Comment (sharonluong) [inline src/main/java/edu/unc/lib/boxc/migration/cdm/services/export/ExportState.java:28]: remember to decide what to do with start time

**Review by sharonluong — APPROVED** [review]
Looks good!

**Pull request merged**